winml/lib/Api/LearningModelSession.cpp

```diff
@@ -28,119 +28,114 @@ static const GUID WINML_PIX_EVAL_CAPTURABLE_WORK_GUID = __uuidof(guid_details::W
-  com_ptr<winmla::IModelProto> model_proto;
+  com_ptr<WinML::IModel> model;
 
   {
     // Lock the model detach/copy since multiple threads can access concurrently
     CWinMLAutoLock lock(&session_creation_lock_);
 
     // Throw if the model has been disposed and is not capable of creating
     // new sessions.
-    auto model = model_.as<winmlp::LearningModel>();
-    WINML_THROW_HR_IF_TRUE_MSG(E_INVALIDARG, model->IsDisposed(),
+    auto model_impl = model_.as<winmlp::LearningModel>();
+    WINML_THROW_HR_IF_TRUE_MSG(E_INVALIDARG, model_impl->IsDisposed(),
                                "The model has been disposed.");
 
-    model_proto.attach(should_close_model
-                      ? model->DetachModelProto()
-                      : model->CopyModelProto());
+    model.attach(should_close_model
+                     ? model_impl->DetachModel()
+                     : model_impl->CloneModel());
   }
 
   // Ensure that the model is runnable on the device
-  com_ptr<winmla::IWinMLAdapter> adapter;
-  WINML_THROW_IF_FAILED(OrtGetWinMLAdapter(adapter.put()));
-  WINML_THROW_IF_FAILED(adapter->EnsureModelDeviceCompatibility(model_, model_proto.get(), device_.as<winmlp::LearningModelDevice>()->GetD3DDeviceCache()->IsFloat16Supported()));
-
-  return model_proto.detach();
+  auto isFloat16Supported = device_.as<winmlp::LearningModelDevice>()->GetD3DDeviceCache()->IsFloat16Supported();
+  if (!isFloat16Supported) {
+    WINML_THROW_IF_FAILED(model->ModelEnsureNoFloat16());
+  }
+  return model.detach();
 }
 
 void LearningModelSession::Initialize() {
   // Begin recording session creation telemetry
   _winmlt::TelemetryEvent session_creation_event(
-    _winmlt::EventCategory::kSessionCreation);
+      _winmlt::EventCategory::kSessionCreation);
   // Get the optimized model proto from the learning model
-  com_ptr<winmla::IModelProto> model_proto; 
-  model_proto.attach(GetOptimizedModel());
+  com_ptr<WinML::IModel> model;
+  model.attach(GetOptimizedModel());
 
   // Create the session builder
   auto device_impl = device_.as<winmlp::LearningModelDevice>();
+  auto model_impl = model_.as<winmlp::LearningModel>();
 
-  com_ptr<winmla::IWinMLAdapter> adapter;
-  WINML_THROW_IF_FAILED(OrtGetWinMLAdapter(adapter.put()));
+  engine_factory_.copy_from(model_impl->GetEngineFactory());
 
-  com_ptr<winmla::IOrtSessionBuilder> session_builder;
-  WINML_THROW_IF_FAILED(adapter->CreateOrtSessionBuilder(
-    device_impl->GetD3DDevice(), 
-    device_impl->GetDeviceQueue(),
-    session_builder.put()));
+  com_ptr<WinML::IEngineBuilder> engine_builder;
+  engine_factory_->CreateEngineBuilder(engine_builder.put());
 
-  Ort::SessionOptions options(nullptr);
-  WINML_THROW_IF_FAILED(session_builder->CreateSessionOptions(options.put()));
+  if (device_impl->IsCpuDevice() == false) {
+    engine_builder->SetD3D12Device(device_impl->GetD3DDevice());
+    engine_builder->SetID3D12CommandQueue(device_impl->GetDeviceQueue());
```

Comment:
nit: If D3D12Device and D3D12CommandQueue need to be set together and never one without the other, why not have a single method something like: "SetD3D12Resources" that take in both d3d12device and d3d12commandqueue as params.

Reply:
sure.

---
In reply to: [370499816](https://github.com/microsoft/onnxruntime/pull/2891#discussion_r370499816) [](ancestors = 370499816)